File: backend-java/src/main/java/com/example/faceattendance/service/AcademicPeriodService.java
package com.example.faceattendance.service;

import com.example.faceattendance.dto.academicperiod.AcademicPeriodResponse;
import com.example.faceattendance.dto.academicperiod.CreateAcademicPeriodRequest;
import com.example.faceattendance.dto.academicperiod.UpdateAcademicPeriodRequest;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

public interface AcademicPeriodService {

    AcademicPeriodResponse create(
            CreateAcademicPeriodRequest request
    );

    AcademicPeriodResponse getById(Long id);

    List<AcademicPeriodResponse> getAll();

    List<AcademicPeriodResponse> getByCourse(
            String course
    );

    AcademicPeriodResponse getActive(

            String course,
            String batch,
            String semester
    );

    @Transactional(readOnly = true)
    Optional<AcademicPeriodResponse> findActive(
            String course,
            String batch,
            String semester);

    AcademicPeriodResponse update(
            Long id,
            UpdateAcademicPeriodRequest request
    );

    void activate(Long id);

    void deactivate(Long id);

    void delete(Long id);
}

File: frontend/src/services/academicPeriodService.js
import api from './api';

const academicPeriodService = {

  getAll: async () => {
    const response = await api.get('/api/academic-periods');
    return response.data;
  },

  getById: async (id) => {
    const response = await api.get(`/api/academic-periods/${id}`);
    return response.data;
  },

  create: async (data) => {
    const response = await api.post(
      '/api/academic-periods',
      data
    );
    return response.data;
  },

  update: async (id, data) => {
    const response = await api.put(
      `/api/academic-periods/${id}`,
      data
    );
    return response.data;
  },

  activate: async (id) => {
    const response = await api.put(
      `/api/academic-periods/${id}/activate`
    );
    return response.data;
  },

  deactivate: async (id) => {
    const response = await api.put(
      `/api/academic-periods/${id}/deactivate`
    );
    return response.data;
  },

  remove: async (id) => {
    const response = await api.delete(
      `/api/academic-periods/${id}`
    );
    return response.data;
  },

};

export default academicPeriodService;